### PR TITLE
Added Ideal and Robust Soliton Distributions

### LINKED ISSUE #20100: Soliton Distributions
I would like to add the Ideal and Robust Soliton Distributions to the stats module. More on this here - 
[Wikipedia](https://en.wikipedia.org/wiki/Soliton_distribution)
[Luby's Research Paper](http://pages.cs.wisc.edu/~suman/courses/740/papers/luby02lt.pdf)

### PR DESCRIPTION
**Pull request by naveensaigit**
#### References to other Issues or PRs
Continues from #20102
Fixes #20100

#### Brief description of what is fixed or changed
The soliton distributions are used in the theory of erasure correcting codes, which use information redundancy to compensate for transmission errors manifesting as missing data. I have implemented both Ideal and Robust Soliton Distributions in the FRV section of stats module

#### Release Notes

* stats
   * Added 2 new distributions in frv_types.

### CODE REVIEW

**Comment by czgdp1807** [comment]
There are two uncovered lines in [https://codecov.io/gh/sympy/sympy/pull/20514/diff](https://codecov.io/gh/sympy/sympy/pull/20514/diff). You can uses `raises` to check if the correct exceptions are thrown.

File: sympy/stats/frv_types.py [hunk 1]
```diff
@@ -66,17 +70,24 @@ def check(density):
         for p in density.values():
             _value_check((p >= 0, p <= 1),
                         "Probability at a point must be between 0 and 1.")
-        _value_check(Eq(sum(density.values()), 1), "Total Probability must be 1.")
+        val = sum(density.values())
+        _value_check(Eq(val, 1) != S.false, "Total Probability must be 1.")
```
Comment (czgdp1807) [inline sympy/stats/frv_types.py:74]: What's the benefit of using this than the original check?

File: sympy/stats/frv_types.py [hunk 2]
```diff
@@ -33,13 +34,16 @@
...
+'RobustSoliton',
 ]
 
-def rv(name, cls, *args):
+def rv(name, cls, *args, **kwargs):
     args = list(map(sympify, args))
     dist = cls(*args)
-    dist.check(*args)
+    if kwargs.pop('check', True):
+        dist.check(*args)
```
Comment (czgdp1807) [inline sympy/stats/frv_types.py:46]: Why this change?

File: sympy/stats/tests/test_finite_rv.py [hunk 3]
```diff
@@ -395,10 +441,16 @@ def test_FiniteRV():
...
-    raises(ValueError, lambda: FiniteRV('F', {1: S.Half, 2: S.Half, 3: S.Half}))
-    raises(ValueError, lambda: FiniteRV('F', {1: S.Half, 2: Rational(-1, 2), 3: S.One}))
+    raises(ValueError, lambda: FiniteRV('F', {1: S.Half, 2: S.Half, 3: S.Half}, check=True))
+    raises(ValueError, lambda: FiniteRV('F', {1: S.Half, 2: Rational(-1, 2), 3: S.One}, check=True))
     raises(ValueError, lambda: FiniteRV('F', {1: S.One, 2: Rational(3, 2), 3: S.Zero,\
-        4: Rational(-1, 2), 5: Rational(-3, 4), 6: Rational(-1, 4)}))
+        4: Rational(-1, 2), 5: Rational(-3, 4), 6: Rational(-1, 4)}, check=True))
+
+    # purposeful invalid pmf but it should not raise since check=False
+    # see test_drv_types.test_ContinuousRV for explanation
```
Comment (czgdp1807) [inline sympy/stats/tests/test_finite_rv.py:450]: Why `test_ContinuousRV` would be in `test_drv_types.py`

File: sympy/stats/frv_types.py (hunk 2 above)
Comment (naveensaigit) [reply sympy/stats/frv_types.py:46]: This change is already in master. I think its showing up because it was not present when I opened the PR initially and now I synchronized with master

File: sympy/stats/frv_types.py (hunk 1 above)
Comment (naveensaigit) [reply sympy/stats/frv_types.py:74]: Same goes for the change here, its already in master

File: sympy/stats/tests/test_finite_rv.py (hunk 3 above)
Comment (naveensaigit) [reply sympy/stats/tests/test_finite_rv.py:450]: I've only added `test_ideal_soliton` and `test_robust_soliton`. Other changes are already part of master. Maybe its a typo and it should be `test_crv_types.test_ContinuousRV` in the comment instead. Let me know and I'll commit the changes if required

**Comment by naveensaigit** [comment]
> There are two uncovered lines in https://codecov.io/gh/sympy/sympy/pull/20514/diff. You can uses `raises` to check if the correct exceptions are thrown.

Ok, I'll do that

File: sympy/stats/tests/test_finite_rv.py (hunk 3 above)
Comment (czgdp1807) [reply sympy/stats/tests/test_finite_rv.py:454]: May be merge `master` into your branch, so that irrelevant changes aren't shown here. You can fix these typos in a different PR.

Comment (naveensaigit) [reply sympy/stats/tests/test_finite_rv.py:454]: OK, I'll do that. Should I do it now? I'm asking because Travis builds seem to taking a very long time. Maybe doing it beforing merging would be better. Or I'll merge `master` and you could cancel the build as it would be redundant. I'll wait for the current build to pass

Comment (czgdp1807) [reply sympy/stats/tests/test_finite_rv.py:454]: You can do it now. This will be merged at least after 24 hours.

Comment (naveensaigit) [reply sympy/stats/tests/test_finite_rv.py]: Done, the diff is clean now

**Comment by czgdp1807** [comment]
LGTM. Will merge tomorrow if no objections raised.
